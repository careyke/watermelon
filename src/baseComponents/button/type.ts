export type ButtonType =
  | "default"
  | "primary"
  | "warning"
  | "danger"
  | "link"
  | "success";
export type ButtonTypeClass = "btnprimary" | "btnwarning" | "btndanger";

export type ButtonShape = "round" | "circle";
export type ButtonShapeClass = "btnround" | "btncircle";

export type ButtonStyle = "icon-text" | "text-icon";

export type ButtonDirection = "row" | "col";
export type ButtonDirectionClass = "btnrow" | "btncol";
