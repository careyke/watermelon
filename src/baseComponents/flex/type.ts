export type IDirection = "row" | "col";
export type IDirectionClassName = "flexrow" | "flexcol";
