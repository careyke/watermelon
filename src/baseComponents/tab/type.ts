export type TTabKey = string | number;

export type TTabNavItem = { tabTitle: string; tabKey: TTabKey };

export type TTabType = 'default' | 'full';
