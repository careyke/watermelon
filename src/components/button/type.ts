export type ButtonType =
  | "default"
  | "primary"
  | "warning"
  | "danger"
  | "link"
  | "success";
export type ButtonTypeClass = "btnprimary" | "btnwarning" | "btndanger";

export type ButtonShape = "round" | "circle";
export type ButtonShapeClass = "btnround" | "btncircle";

export type ButtonStyle = "icon-text" | "text-icon";
