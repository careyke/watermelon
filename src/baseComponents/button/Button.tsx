import React, { FC, MouseEventHandler, CSSProperties } from 'react';
import classnames from 'classnames';
import styles from './button.less';
import { ButtonType, ButtonShape, ButtonTypeClass, ButtonStyle, ButtonShapeClass, ButtonDirection, ButtonDirectionClass } from './type';

interface IButtonProps {
  type?: ButtonType,
  shape?: ButtonShape,
  icon?: string;
  buttonStyle?: ButtonStyle;
  onClick?: MouseEventHandler;
  style?: CSSProperties;
  direction?: ButtonDirection
}

export const Button: FC<IButtonProps> = (props) => {
  const { children, type = 'default', shape, buttonStyle = 'icon-text', onClick, icon, direction = 'row' } = props;
  const typeClassName = `btn${type}` as ButtonTypeClass;
  const shapeClassName = `btn${shape}` as ButtonShapeClass;
  const directionClassName = `btn${direction}` as ButtonDirectionClass;
  const buttonClass = classnames({
    [styles.button]: true,
    [styles[typeClassName]]: type,
    [styles[shapeClassName]]: shape,
    [styles[directionClassName]]: direction
  });
  const iconClass = classnames({
    [styles.icon]: true,
    [`iconfont icon-${icon}`]: icon
  });
  const textClass = classnames({
    [styles.btnicontext]: icon && buttonStyle === 'icon-text',
    [styles.btntexticon]: icon && buttonStyle === 'text-icon'
  });

  return (
    <div className={buttonClass} onClick={onClick}>
      {buttonStyle === 'icon-text' && icon ? <span className={iconClass}></span> : null}
      {shape === 'circle' && icon ? null : <span className={textClass}>{children}</span>}
      {buttonStyle === 'text-icon' && icon ? <span className={iconClass}></span> : null}
    </div>
  );
}