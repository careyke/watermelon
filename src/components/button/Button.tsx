import React, { useState } from 'react';
import classnames from 'classnames';
import { IButtonProps, ButtonType, ButtonShape, ButtonShowStyle } from './type';
import styles from './button.less';

export default function Button(props: IButtonProps): React.ReactElement {
  const { type = 'default', shape, loading = false, showStyle = 'icon-text', icon } = props;
  let iconClass = classnames({
    [styles['btn-icon']]: true,
    'iconfont': true,
    ['icon-' + icon]: true
  })
  let btnClass = classnames({
    [styles['btn-base']]: true,
    [styles['btn-default']]: type === ButtonType.Default,
    [styles['btn-primary']]: type === ButtonType.Primary,
    [styles['btn-danger']]: type === ButtonType.Danger,
    [styles['btn-link']]: type === ButtonType.Link,
    [styles['btn-round']]: shape === ButtonShape.Round,
    [styles['btn-circle']]: shape === ButtonShape.Circle,
  })
  let btnTextClass = classnames({
    [styles['btn-text']]: true,
    [styles['btn-icon-text']]: icon && showStyle === ButtonShowStyle.IconText,
    [styles['btn-text-icon']]: icon && showStyle === ButtonShowStyle.TextIcon,
  })
  // const [isLoading, setLoading] = useState(loading);

  return (
    <div className={btnClass}>
      {
        showStyle === ButtonShowStyle.IconText && icon ? <span className={iconClass} ></span> : null
      }
      {shape !== ButtonShape.Circle ? <span className={btnTextClass}>{props.children}</span> : null}
      {
        showStyle === ButtonShowStyle.TextIcon && icon ? <span className={iconClass} ></span> : null
      }
    </div>
  )
}