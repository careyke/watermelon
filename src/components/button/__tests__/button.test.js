import React from "react";
import renderer from "react-test-renderer";
import { render, mount, shallow } from "enzyme";
import Button from "../Button";
import btnStyles from "../button.less";
import { ButtonType } from "../type";

describe("Button test", () => {
  it("renders correctly", () => {
    const wrapper = render(<Button>Default</Button>);
    expect(wrapper).toMatchSnapshot();
  });
  it("renders correctly with icon left", () => {
    const wrapper = render(
      <Button icon="search" showStyle="icon-text">
        Default
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });
  it("render correctly with icon right", () => {
    const wrapper = render(
      <Button icon="search" showStyle="text-icon">
        Default
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });
  it("render correctly without text", () => {
    const wrapper = render(
      <Button icon="search" shape="circle">
        Default
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });
  it("mount correctly", () => {
    const wrapper = mount(<Button>Default</Button>);
    expect(() => wrapper).not.toThrow();
  });
  it("render with correctly class", () => {
    const typeKeys = Object.keys(ButtonType);
    const len = typeKeys.length;
    const getRandType = () => {
      let index = ~~(Math.random() * len);
      return ButtonType[typeKeys[index]];
    };
    const currentType = getRandType();
    const wrapper = shallow(<Button type={currentType} shape={'circle'} >Default</Button>);
    expect(wrapper.find(".btn-base").hasClass(`btn-${currentType}`)).toBe(true);
    expect(wrapper.find(".btn-base").hasClass('btn-circle')).toBe(true);
  });
});
