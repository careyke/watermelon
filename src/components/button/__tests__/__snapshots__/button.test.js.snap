// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button test render correctly with icon right 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "btn-base btn-default",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "btn-text btn-text-icon",
        },
        "children": Array [
          Object {
            "data": "Default",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "btn-icon iconfont icon-search",
          },
          "children": Array [],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "btn-icon iconfont icon-search",
        },
        "children": Array [],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "btn-text btn-text-icon",
          },
          "children": Array [
            Object {
              "data": "Default",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Button test render correctly without text 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "btn-base btn-default btn-circle",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "btn-icon iconfont icon-search",
        },
        "children": Array [],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Button test renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "btn-base btn-default",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "btn-text",
        },
        "children": Array [
          Object {
            "data": "Default",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Button test renders correctly with icon left 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "btn-base btn-default",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "btn-icon iconfont icon-search",
        },
        "children": Array [],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "btn-text btn-icon-text",
          },
          "children": Array [
            Object {
              "data": "Default",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "btn-text btn-icon-text",
        },
        "children": Array [
          Object {
            "data": "Default",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "btn-icon iconfont icon-search",
          },
          "children": Array [],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
